src/Sulu/Bundle/MediaBundle/Collection/Manager/CollectionManager.php

```diff
@@ -797,4 +841,54 @@ private function getCollectionMeta(CollectionInterface $collection, ?string $loc
 
         return $meta;
     }
+
+    /**
+     * @param array<array{id: int, resourceKey: string, depth: int}> $resources
+     *
+     * @return array<int, array<array{id: int, resourceKey: string}>>
+     */
+    private function groupResourcesByDepth(array $resources): array
+    {
+        $grouped = [];
+
+        foreach ($resources as $resource) {
+            $depth = $resource['depth'];
+            unset($resource['depth']);
+
+            if (!isset($grouped[$depth])) {
+                $grouped[$depth] = [];
+            }
+
+            $grouped[$depth][] = $resource;
+        }
+
+        \krsort($grouped);
+
+        return \array_values($grouped);
+    }
+
+    /**
+     * @return array<int, array<array{id: int, resourceKey: string}>>
+     */
+    private function findAllChildResourcesByRootCategory(int $rootCollectionId): array
+    {
+        $childCollections = $this->collectionRepository->findChildCollectionResourcesOfRootCollection($rootCollectionId);
+
+        $collectionIds = \array_merge([$rootCollectionId], \array_column($childCollections, 'id'));
+        $childMedia = $this->mediaRepository->findMediaResourcesOfCollections($collectionIds);
+
+        $result = \array_merge($childCollections, $childMedia);
+
+        return $this->groupResourcesByDepth($result);
+    }
+
+    private function countAllChildrenByRootCategory(int $id): int
+    {
+        $countChildCollections = $this->collectionRepository->countChildCollectionsOfRootCollection($id);
+
+        $childCollectionIds = \array_merge([$id], $this->collectionRepository->findChildCollectionIdsOfRootCollection($id));
```

Comment:
I don't see a need for this method it should just give the `rootColectionId` into the `MediaRepository` and use the nested tree to get all children.